src/render/src/encoder.rs

```diff
@@ -150,456 +178,436 @@ impl<B: Backend> AccessInfo<B> {
-    pub fn flush(self, queue: &mut GraphicsQueue<B>) -> SubmissionResult<()> {
-        self.synced_flush(queue, &[], &[], None)
-    }
-
-    /// Submits the commands in the internal `CommandBuffer` to the GPU, so they can
-    /// be executed.
-    pub fn synced_flush(self,
-                        queue: &mut GraphicsQueue<B>,
-                        wait_semaphores: &[&handle::Semaphore<B>],
-                        signal_semaphores: &[&handle::Semaphore<B>],
-                        fence: Option<&handle::Fence<B>>) -> SubmissionResult<()> {
-        self.finish().synced_flush(queue, wait_semaphores, signal_semaphores, fence)
-    }
-
-    /// Copy part of a buffer to another
-    pub fn copy_buffer<T: Pod>(&mut self, src: &handle::Buffer<B, T>, dst: &handle::Buffer<B, T>,
-                               src_offset: usize, dst_offset: usize, size: usize) -> CopyBufferResult {
-        if !src.get_info().bind.contains(memory::TRANSFER_SRC) {
-            return Err(CopyError::NoSrcBindFlag);
-        }
-        if !dst.get_info().bind.contains(memory::TRANSFER_DST) {
-            return Err(CopyError::NoDstBindFlag);
+    pub fn init_resources(&mut self, tokens: Vec<InitToken<B>>) {
+        let mut barriers = Vec::new();
+        for token in &tokens {
+            match token.handle {
+                handle::Any::Image(ref image) =>
+                    barriers.push(self.init_image(image)),
+                _ => {}
+            }
         }
+        let stage_transition = self.pipeline_stage..self.pipeline_stage;
+        self.buffer.pipeline_barrier(stage_transition, &barriers[..]);
+    }
 
-        let size_bytes = mem::size_of::<T>() * size;
-        let src_offset_bytes = mem::size_of::<T>() * src_offset;
-        let src_copy_end = src_offset_bytes + size_bytes;
-        if src_copy_end > src.get_info().size {
-            return Err(CopyError::OutOfSrcBounds {
-                size: src.get_info().size,
-                copy_end: src_copy_end,
-            });
-        }
-        let dst_offset_bytes = mem::size_of::<T>() * dst_offset;
-        let dst_copy_end = dst_offset_bytes + size_bytes;
-        if dst_copy_end > dst.get_info().size {
-            return Err(CopyError::OutOfDstBounds {
-                size: dst.get_info().size,
-                copy_end: dst_copy_end,
-            });
-        }
-        if src == dst &&
-           src_offset_bytes < dst_copy_end &&
-           dst_offset_bytes < src_copy_end
-        {
-            return Err(CopyError::Overlap {
-                src_offset: src_offset_bytes,
-                dst_offset: dst_offset_bytes,
-                size: size_bytes,
-            });
+    fn init_image<'b>(&mut self, image: &'b handle::raw::Image<B>)
+        -> core::memory::Barrier<'b, B>
+    {
+        let creation_state = (core::image::Access::empty(), ImageLayout::Undefined);
+        let levels = image.info().kind.get_num_levels();
```

Comment:
Looks like this does not give the actual image levels, should I use `image.info().mip_levels` instead ?.

Reply:
I'd expect them to match, plus minus the differences between 0 and 1

Reply:
The way I see it, `kind.get_num_levels()` is how many mip levels you should have if you had mip mapping while `info().mip_levels` is how many mip levels you really have (so it is valid even if you do not have mip mapping).

Reply:
I think you just need to remove `Info::mip_levels` completely. Image `Kind` is the full description of what an image has in terms of dimensions/levels.

Reply:
Do we agree that:

1. You can have images with arbitrary mip levels (1 most likely, like for backbuffers).
2. `kind.get_num_levels()` will not return this number, but rather `(1..).find(|level| dominant>>level <= 1).unwrap()`, which is like the maximum number of mip levels you can have.

And that we need the actual level count and not the maximum level count ?

Reply:
@Bastacyclop Sorry, I got confused. Used the `get_num_layers` lately and assumed `get_num_levels` does a similar thing. You are correct, it doesn't return the actually stored mipmap count, this method can only be used for estimation during the image creation, and is rather pointless afterwards.

So in your case using `image.info().mip_levels` is the way to proceed.